add: yamux (#268)
libp2p/docs: content/concepts/multiplex/yamux.md

---
title: "Yamux"
description: "yamux is a multiplexing protocol designed by Hashicorp."
weight: 180
---

## What is Yamux?

[Yamux](https://github.com/hashicorp/yamux) is a multiplexing protocol designed
by [Hashicorp](https://www.hashicorp.com/) for Go. It is built over
[SCTP (Stream Control Transmission Protocol)](https://en.wikipedia.org/wiki/Stream_Control_Transmission_Protocol), which is a transport layer protocol
that provides reliable, message-oriented communication with congestion control.
Yamux provides a high-level interface for working with SCTP streams. It is inspired

> by [SPDY](https://en.wikipedia.org/wiki/SPDY), but is not interoperable with it.

Reviewer: It's built over TCP, not SCTP.

Yamux offers more sophisticated flow control than [mplex](mplex), and can scale

> to thousands of multiplexed streams over a single connection.

Reviewer: mplex can also handle 1000s of streams. We should say that it does support backpressure, and maybe explain how backpressure works (the receiver allows sending of stream data up to a certain offset, and that offset is increased as data is processed by the application), and why you need backpressure (application might be slow processing the data, and we don't want to buffer an unlimited amount of data).

---
title: "Yamux"
description: "yamux is a multiplexing protocol designed by Hashicorp."
weight: 180
---

## What is Yamux?

[Yamux](https://github.com/hashicorp/yamux) (Yet another Multiplexer)
is a powerful stream multiplexer used in libp2p. It was

> developed by Hashicorp for Go and built on top of TCP. Yamux

Reviewer: ```suggestion
initially developed by Hashicorp for Go, and is now implemented in Rust, JavaScript and other languages.
```

> enables multiple parallel streams on a single TCP connection
> and provides a high-level interface similar to SPDY, however
> it is not compatible with it.

Reviewer: ```suggestion
enables multiple parallel streams on a single TCP connection.
The design was inspired by SPDY (which later became the basis for HTTP/2), however
it is not compatible with it.
```

---
title: "Yamux"
description: "yamux is a multiplexing protocol designed by Hashicorp."
weight: 180
---

> ## What is Yamux?

Reviewer: I suggest adding an introduction sentence of what purpose Yamux serves in the libp2p stack.

[Yamux](https://github.com/hashicorp/yamux) (Yet another Multiplexer) is a multiplexing
protocol initially designed by [Hashicorp](https://www.hashicorp.com/) for Go, built over
TCP. Yamux provides a high-level interface inspired
by [SPDY](https://en.wikipedia.org/wiki/SPDY), but it is not interoperable with it.

> 

Reviewer: First we should say that Yamux allows us to have multiple streams in parallel on a single TCP connection.

Yamux supports flow control through backpressure, which is a mechanism that helps to
prevent data from being sent faster than it can be processed.

One way to achieve backpressure is for the receiver to specify an offset
to which the sender can send data. This offset increases as the receiver consumes

> the data, allowing the sender to send more data. This helps prevent the
> sender from sending more data than the receiver can handle, leading to buffering
> and potentially causing the application to crash. This is especially important in

Reviewer: ```suggestion
the data, allowing the sender to send more data. This helps prevent the
sender from sending data faster than the receiver process it. This is especially important in
```

applications where the receiver might be slower at processing the data, such as when
the receiver is a low-powered device or when the data is complex and requires a lot of

> processing.

Reviewer: Should this be one paragraph?

---
title: "Yamux"
description: "yamux is a multiplexing protocol designed by Hashicorp."
weight: 180
---

## What is Yamux?

[Yamux](https://github.com/hashicorp/yamux) is a multiplexing protocol designed
by [Hashicorp](https://www.hashicorp.com/) for Go, built over TCP.
Yamux provides a high-level interface for working with SCTP streams. It is inspired

> by [SPDY](https://en.wikipedia.org/wiki/SPDY), but it is not interoperable with it.

Reviewer: Yamux doesn't use SCTP.

Yamux offers more sophisticated flow control than [mplex](mplex) while also supporting
backpressure, which is a mechanism that helps to prevent data from being sent faster
then it can be processed.

Backpressure allows the receiver of a data stream to specify an offset

> to which the sender can send data. This offset increases as the receiver process the data,

Reviewer: This paragraph should be rephrased. Backpressure in general is just a mechanism to prevent one peer to overwhelm a peer that's slow consuming the data.
Sending the offset is _one_ concrete option to implement backpressure.

> to which the sender can send data. This offset increases as the receiver process

Reviewer: ```suggestion
to which the sender can send data. This offset increases as the receiver consumes
```

allowing the sender to send more data. This helps prevent the
sender from sending more data than the receiver can handle, leading to buffering
and potentially causing the application to crash. This is especially important in
applications where the receiver might be slower at processing the data, such as when the
receiver is a low-powered device or when the data is complex and requires a lot of processing.

{{< alert icon="💡" context="note">}}
**Yamux should be used over mplex in libp2p**. It is well-suited for applications that require fast,

> reliable data transfer, such as real-time communication or streaming media as it is optimized for low-latency, high-bandwidth environments.

Reviewer: Not really. When you're streaming media, you typically want an unreliable transport.
